File: 13-05-2024/Number of Good Components.js
//{ Driver Code Starts

process.stdin.resume();
process.stdin.setEncoding('utf-8');

let inputString = '';
let currentLine = 0;

process.stdin.on('data', inputStdin => { inputString += inputStdin; });

process.stdin.on('end', _ => {
    inputString =
        inputString.trim().split("\n").map(string => { return string.trim(); });

    main();
});

function readLine() { return inputString[currentLine++]; }

function main() {
    let t = parseInt(readLine());
    let i = 0;
    for (; i < t; i++) {

        let e = parseInt(readLine());

        let v = parseInt(readLine());

        let edges = [];
        for (let i = 0; i < e; i++) {
            let a = readLine().trim().split(' ').map((x) => parseFloat(x));
            edges.push(a);
        }

        let obj = new Solution();
        let res = obj.findNumberOfGoodComponent(e, v, edges);
        console.log(res);
    }
}

// } Driver Code Ends


//Your Code Start
class Solution {
    /**
    * @param {number} e
    * @param {number} v
    * @param {number[][]} edges
    * @returns {number}
    */
    findNumberOfGoodComponent(e, v, edges) {
        function isConnected(check, visited, adj) {
            const checkList = [...adj[check], check].sort((a, b) => a - b);
            for (const i of adj[check]) {
                visited[i] = true;
                const temp = [...adj[i], i].sort((a, b) => a - b);
                if (checkList.join(',') !== temp.join(',')) return false;
            }
            return true;
        }

        let ans = 0;
        const visited = new Array(v + 1).fill(false);
        const adj = Array.from({ length: v + 1 }, () => []);
        
        for (const [start, end] of edges) {
            adj[start].push(end);
            adj[end].push(start);
        }
        
        for (let i = 1; i <= v; i++) {
            if (!visited[i]) {
                const isConnectedComponent = isConnected(i, visited, adj);
                if (isConnectedComponent) ans++;
            } 
        }
        
        return ans;
    }
}
//Your Code End

File: 13-05-2024/Number of Good Components.java
//{ Driver Code Starts
import java.io.*;
import java.util.*;

class IntMatrix {
    public static int[][] input(BufferedReader br, int n, int m) throws IOException {
        int[][] mat = new int[n][];

        for (int i = 0; i < n; i++) {
            String[] s = br.readLine().trim().split(" ");
            mat[i] = new int[s.length];
            for (int j = 0; j < s.length; j++) mat[i][j] = Integer.parseInt(s[j]);
        }

        return mat;
    }

    public static void print(int[][] m) {
        for (var a : m) {
            for (int e : a) System.out.print(e + " ");
            System.out.println();
        }
    }

    public static void print(ArrayList<ArrayList<Integer>> m) {
        for (var a : m) {
            for (int e : a) System.out.print(e + " ");
            System.out.println();
        }
    }
}

class GFG {
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        int t;
        t = Integer.parseInt(br.readLine());
        while (t-- > 0) {

            int e;
            e = Integer.parseInt(br.readLine());

            int v;
            v = Integer.parseInt(br.readLine());

            int[][] edges = IntMatrix.input(br, e, 2);

            Solution obj = new Solution();
            int res = obj.findNumberOfGoodComponent(e, v, edges);

            System.out.println(res);
        }
    }
}

// } Driver Code Ends



//Your Code Start
class Solution {
    public static boolean isConnected(int check, boolean[] visited, List<List<Integer>> adj) {
        List<Integer> checkList = new ArrayList<>(adj.get(check));
        checkList.add(check);
        Collections.sort(checkList);
        
        for (int i : adj.get(check)) {
            visited[i] = true;
            List<Integer> temp = new ArrayList<>(adj.get(i));
            temp.add(i);
            Collections.sort(temp);
            if (!checkList.equals(temp)) return false;
        }
        return true;
    }
    
    public static int findNumberOfGoodComponent(int e, int v, int[][] edges) {
        int ans = 0;
        boolean[] visited = new boolean[v + 1];
        List<List<Integer>> adj = new ArrayList<>(v + 1);
        for (int i = 0; i <= v; i++) {
            adj.add(new ArrayList<>());
        }
        
        for (int[] edge : edges) {
            int start = edge[0];
            int end = edge[1];
            
            adj.get(start).add(end);
            adj.get(end).add(start);
        }
        
        for (int i = 1; i <= v; i++) {
            if (!visited[i]) {
                boolean isConnectedComponent = isConnected(i, visited, adj);
                if (isConnectedComponent) ans++;
            } 
        }
        
        return ans;
    }
}
//Your Code End
